### PR TITLE
Signed callbacks

### PR DESCRIPTION
**Pull request by niati**
## Description

Added new documentation on signed-callbacks via Messages API.

[Review App](https://nexmo-develo-signed-cal-u4agq1.herokuapp.com/messages/concepts/signed-webhooks)

### CODE REVIEW

**Comment by niati** [comment]
@tbedford I committed only 4 files for this PR but the 'Files Changed' tab includes 13 files changed. Not sure why that is happening. Can you take a look and let me know what you think?

**Comment by tbedford** [comment]
@niati - yes it looks like you had some modified files that have been picked up as part of this PR. However, those changes don't look harmful - I would just check the links and make sure they're OK.

**Comment by lornajane** [comment]
I think it's this commit that shouldn't be here https://github.com/Nexmo/nexmo-developer/pull/2848/commits/35f7308722aff6ada0f094635be36ba752ef64a0 - I can remove it from the branch if that helps? (I just dropped by to read the docs on the message signing stuff since I'm very interested in this area)

**Comment by niati** [comment]
@lornajane That would be great. The build is failing and I'm wondering if that's why.

**Comment by tbedford** [comment]
@niati - build is failing because you are not pointing at latest master of the code snippet submodule - it's not because of the modifications that crept in. I can show you how to fix if you want. Once it's building I can add my review.

**Comment by niati** [comment]
@tbedford That would be great, Tony. I think we should still back-out the commit Lorna mentioned. That includes the unnecessary links to the tutorials you re-worked a few weeks ago. I'm pretty sure those links will break things.

**Comment by tbedford** [comment]
OK let's let @lornajane make her change first.

**Comment by tbedford** [comment]
@lornajane - thanks! 
----
@niati - there are some spelling mistakes to fix. Variable names and such should be in back ticks. 

```
Everything's up-to-date. Nothing to do
    _documentation/en/messages/concepts/signed-webhooks.md
        9 | re secret associated with the apikey included in the JWT claims. 
       19 |  tampered with by comparing a sha_256 hash of the payload to the pa 
       19 | 56 hash of the payload to the payload_hash field found in the JWT claims 
       61 | `payload_hash` | A sha256 hash of the request payload.  
       62 | `api_key` | a1b2c3d | The API key associated with 
       63 | `application_id` | appId1 | (Optional) The id of the ap 

>> 6 spelling errors found in 362 files
```

**Comment by lornajane** [comment]
OK, try that! I rebased onto current master and put your two commits on, the diff looks right to me now. The build failure is for some spellings, I'm sure you can sort that bit out! Thanks for these docs, I had questions about this feature and this helps :+1:

**Comment by lornajane** [comment]
There is a bad merge in the history here. @niati I think you merged your previous changes into the branch

**Comment by niati** [comment]
@lornajane I think it happened when I made the spelling corrections. I'm not sure why. I went back and manually updated the extra files that were merged to match what's in master.

**Comment by tbedford** [comment]
@niati - you fixed the spelling issues by adding a bunch of variables to the spelling file. This is not recommended. Please see comment above - variables should have backticks around them - this makes sure the spell checker skips over them. Thanks.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+---
+title: Signed webhooks
+description: Verify the source of incoming requests via the Messages API.
+navigation_weight: 4
+---
+
+# Signing webhooks
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:7]: The title in metadata is "Signed webhooks" but h1 is "Signing webhooks". Is this topic about signing webhooks, signed webhooks, or verifying signed webhooks? Some clarification is required.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+Signed webhooks provide a method for your application to verify a request is coming from Vonage and its payload has not been tampered with during transit. When receiving a request, the incoming webhook will include a JWT token in the authorization header which is signed with your signature secret. The secret used to sign the request corresponds to the signature secret associated with the `api_key` included in the JWT claims.
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:9]: Are signed webhooks something I need to enable in Dashboard for Messages API, or do I need to request enablement? Does it work for Dispatch API?

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+You use a signature to:
+
+* Verify that a request originates from a trusted source
+* Ensure that the message has not been tampered with en-route
+* Defend against interception and later replay
+
+## Validate the signature on incoming messages
+
+Webhooks for both inbound messages and message status will include a JWT in the authorization header. Use the API key included in the JWT claims to identify which of your signature secrets has been used to sign the request. Once you have verified the authenticity of the request, you can verify the request payload has not been tampered with by comparing a SHA-256 hash of the payload to the `payload_hash` field found in the JWT claims. If they do not match, then the payload has been tampered with during transit.
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:19]: The section heading is "validating signature on incoming messages" - I thought we were validating webhooks? Is there a difference? There seem to be two parts to validation 1) Verifying the request 2) Verifying the payload has not been tampered with in-transit. Perhaps a numbered list would help make that stand out. How do I verify the authenticity of the request. The verifying of the payload is clear though.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+In the rare case of an Internal Error, it is possible that the Callback Service will send an unsigned callback. By returning an HTTP 5xx response, a retry will be triggered giving the system time to resolve the error and sign future callbacks.
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:21]: Why is 'Internal Error' capitalized? Why is 'Callback Service' capitalized? There may be a reason.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+The code example below shows how to verify a webhook signature. We recommend using HTTPS protocol as it ensures that the request and response are encrypted on the both (client and server) the ends.
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:23]: Who is "we"?

Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:23]: > 'ensures that the request and response are encrypted on the both (client and server) the ends.' 

I would clarify the wording which looks wrong. Perhaps:

ensures that the request and response are encrypted on both the client and server.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+```code_snippets
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:25]: The code could perhaps make a little use of whitespace. I would normally put a blank like before a function definition for example. Totally a personal thing, but the code looks a bit cramped to me.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+source: '_examples/messages/signed-webhooks'
+```
+
+## Sample Signed JWT
+
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:30]: There's no context text. We should add some explanatory text to clarify what this thing following actually is.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+```json
+{
+  "alg": "HS256",
+  "typ": "JWT",
+}
+// body
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:36]: We used body here, but payload in the following table. Also there should probably be a comment for // header too.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  "iat": 1587494962,
+  "jti": "c5ba8f24-1a14-4c10-bfdf-3fbe8ce511b5",
+  "iss": "Vonage",
+  "payload_hash" : "d6c0e74b5857df20e3b7e51b30c0c2a40ec73a77879b6f074ddc7a2317dd031b",
+  "api_key": "a1b2c3d",
+  "application_id": "aaaaaaaa-bbbb-cccc-dddd-0123456789ab"
+}
+```
+
+### Header
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:47]: Header of what?

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:48]: We need some context text prior to the table to explain what this actually is.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+Header | Value
+-- | --
+`alg` | HS256
+`typ` | JWT
+
+### Payload
+
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:55]: Also add some context text here to explain what we have in the table.

File: _examples/messages/signed-webhooks/node.yml [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+dependencies:
+    - jsonwebtoken
+code:
+    source: .repos/nexmo/nexmo-node-code-snippets/messages/signed-webhooks/verify-signed-webhook.js
+    from_line: 11
+    to_line: 33
+client:
+    source: .repos/nexmo/nexmo-node-code-snippets/messages/signed-webhooks/verify-signed-webhook.js
+    from_line: 2
+    to_line: 10
```
Comment (tbedford) [inline _examples/messages/signed-webhooks/node.yml:11]: Need a carriage return here according to GitHub.

**Comment by tbedford** [comment]
Fixed merge conflict.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
+---
+title: Signed webhooks
+description: Verify the source of incoming requests via the Messages API.
+navigation_weight: 4
+---
+
+# Signed webhooks
+
+Signed webhooks are supported by both the Messages and Dispatch APIs and are enabled by default for Messages API traffic. They provide a method for your application to verify a request is coming from Vonage and its payload has not been tampered with during transit. When receiving a request, the incoming webhook will include a JWT token in the authorization header which is signed with your signature secret.
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:9]: "Signed webhooks are supported by both the Messages and Dispatch APIs and are enabled by default for Messages API traffic." I would reword this slightly to:

Signed webhooks are supported by both the Messages and Dispatch APIs and are enabled by default.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+You use a signature to:
+
+* Verify that a request originates from a trusted source
+* Ensure that the message has not been tampered with en-route
+* Defend against interception and later replay
+
+## Validating signed webhooks
+
+There is two parts to validating signed webhooks:
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:19]: is -> are

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+navigation_weight: 4
+---
+
+# Signed webhooks
+
+Signed webhooks are supported by both the Messages and Dispatch APIs and are enabled by default for Messages API traffic. They provide a method for your application to verify a request is coming from Vonage and its payload has not been tampered with during transit. When receiving a request, the incoming webhook will include a JWT token in the authorization header which is signed with your signature secret.
+
+You use a signature to:
+
+* Verify that a request originates from a trusted source
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:13]: I wasn't sure what we are checking for when we say 'trusted source' - what's the thing we are checking - is it API key or something else?

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+* Ensure that the message has not been tampered with en-route
+* Defend against interception and later replay
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:15]: How?

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+## Validating signed webhooks
+
+There is two parts to validating signed webhooks:
+
+1. Verify the request
+2. Verify the payload (optional)
+
+### Verify the request
+
+Webhooks for both inbound messages and message status will include a JWT in the authorization header. Use the API key included in the JWT claims to identify which of your signature secrets has been used to sign the request. The secret used to sign the request corresponds to the signature secret associated with the `api_key` included in the JWT claims. You can find your signature secret on the [Dashboard](https://dashboard.nexmo.com/settings).
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:26]: I'm not quite sure how this paragraph "verifies the request"? It feels like there should be an additional step e.g. identify the correct signature key and then...

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+### Verify the payload has not been tampered with in transit
+
+Once you have verified the authenticity of the request, you may optionally verify the request payload has not been tampered with by comparing a SHA-256 hash of the payload to the `payload_hash` field found in the JWT claims. If they do not match, then the payload has been tampered with during transit.
+
+> **NOTE:** In the rare case of an internal error, it is possible that the callback service will send an unsigned callback. By returning an HTTP 5xx response, a retry will be triggered giving the system time to resolve the error and sign future callbacks.
+
+The code example below shows how to verify a webhook signature. It is recommended you use HTTPS protocol as it ensures that the request and response are encrypted on both the client and server ends.
+
+```code_snippets
+source: '_examples/messages/signed-webhooks'
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:37]: Code snippet is not displaying for me in the review app.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+The contents of the signed JWT header are described in the table below.
+
+Header | Value
+-- | --
+`alg` | HS256
+`typ` | JWT
+
+### Signed JWT Payload
+
+The contents of the signed JWT payload are described in the table below using the values included in the sample signed JWT provided above.
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:70]: Minor point, but generally we avoid the use of 'above' and 'below', as depending on format/device etc. the text may not be. You could say instead 'in the following table, '. Also, another minor point, generally before anything that breaks the text e.g. table, code block we use a colon, rather than full stop. So instead say:

"The contents of the signed JWT payload are described in the following table, using the values included in the sample signed JWT show previously:" - you could actually make the later a link too.

File: _documentation/en/messages/concepts/signed-webhooks.md (hunk 18 above)
Comment (tbedford) [reply _documentation/en/messages/concepts/signed-webhooks.md:37]: No - it's fine.

File: _documentation/en/messages/concepts/signed-webhooks.md [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+1. Verify the request
+2. Verify the payload (optional)
+
+### Verify the request
+
+Webhooks for both inbound messages and message status will include a JWT in the authorization header. Use the API key included in the JWT claims to identify which of your signature secrets has been used to sign the request. The secret used to sign the request corresponds to the signature secret associated with the `api_key` included in the JWT claims. You can identify your signature secret on the [Dashboard](https://dashboard.nexmo.com/settings).
+
+### Verify the payload has not been tampered with in transit
+
+Once you have verified the authenticity of the request, you may optionally verify the request payload has not been tampered with by comparing a SHA-256 hash of the payload to the `payload_hash` field found in the JWT claims. If they do not match, then the payload has been tampered with during transit.
```
Comment (tbedford) [inline _documentation/en/messages/concepts/signed-webhooks.md:30]: We do say checking payload is optional, but don't say anything on why we might want to. My understanding is you only need to do this if you're *not* using HTTPS. Whatever, the reason we should say something about it.

Comment (lornajane) [reply _documentation/en/messages/concepts/signed-webhooks.md:30]: I think we should recommend that everyone checks payloads, every time. Not everyone will, and it doesn't break anything if you don't, but let's encourage everyone to use the security features we provide!

Comment (mheap) [reply _documentation/en/messages/concepts/signed-webhooks.md:30]: I disagree @lornajane 

Checking the payload adds a lot of additional noise and might put people off

Verifying the JWT with the shared secret is enough if you're using HTTPS as TLS will prevent MITM attacks (which is the only reason to validate the payload hash)

**Comment by slorello89** [comment]
@niati - fyi this feature is now live so it can be merged when ready.
